Accomodate JSON, Array, escaped string params

richmolj opened the pull request:
* The type `hash` can now be queried with a string of JSON, and will get
converted to a ruby hash.
* The type `array` can now pass a JSON array in the query string.
* Strings with special characters/commas can now be wrapped in
{{curlies}} to be treated straight-up.
* If an attribute exists, but a filter defines a different type, the
filter type will now be honored when filtering.

See specs for more.

richmolj commented:
@wadetandy as discussed

wadetandy commented on lib/graphiti/types.rb:
```diff
@@ -69,7 +69,8 @@ def self.create(primitive, &blk)
 
     PresentParamsHash = create(::Hash) do |input|
-      Dry::Types['params.hash'][input].deep_symbolize_keys
+      input = JSON.parse(input) if input.is_a?(String)
```
assuming that DryTypes already rescues and throws some sort of helpful error if `JSON.parse` blows up?

wadetandy commented on lib/graphiti/scoping/filter.rb:
```diff
@@ -115,5 +95,90 @@ def coerce_types(name, value)
+      # merge everything back together into an array
+      value = Array(value.split(',')) + quotes
+      # remove any blanks that are left
+      value.reject! { |v| v.length.zero? }
```
Just checking: this means our filtering system does not allow me filter records where a string field is blank or nil.  Is that desired behavior?  How can I achieve this without adding a custom filter?

wadetandy commented on lib/graphiti/scoping/filter.rb:
```diff
@@ -115,5 +95,90 @@ def coerce_types(name, value)
+      [value, operator]
+    end
+
+    # foo,bar,baz becomes ["foo", "bar", "baz"]
```
Since we are commenting on this behavior, I'd add that arrays are used as-is.

wadetandy commented on lib/graphiti/scoping/filter.rb:
```diff
@@ -115,5 +95,90 @@ def coerce_types(name, value)
+    # {{{ "id": 1 }}} becomes { 'id' => 1 }
+    def parse_string_value(value)
+      if (arr = value.scan(/\[.*?\]/)).present?
+        value = arr.map { |json| JSON.parse(json) }
```
Should figure out a friendly error when `JSON.parse` blows up.  At the very least lets add a TODO

wadetandy commented on lib/graphiti/scoping/filter.rb:
```diff
@@ -115,5 +95,90 @@ def coerce_types(name, value)
+      end
+    end
+
+    def validate_whitelist(resource, filter, values)
```
DHH opened [this issue](https://github.com/rails/rails/issues/33677) on Rails a couple days ago (and a [follow up](https://github.com/rails/rails/pull/33681) PR was merged). Since we are in this code anyway, I think we should match rails' terminology.

The linked docs PR seems to have settled on `permitted` and `restricted`.

richmolj commented:
@wadetandy take a look now. Added custom error for when JSON.parse blows up, additional specs for string array json, and changed all whitelist/blacklist to allowlist/denylist
